File: apps/web/lib/dashboard/posture.ts
import type { FindingSeverity } from '@/lib/feed/findings'

/**
 * The dashboard headline (ENT-77): a single Green / Amber / Red band that tells
 * a founder in two seconds whether their compliance is in trouble.
 *
 * The rule is the AC, so it lives here as a pure function over already-loaded
 * data — exhaustively unit-tested, and reused by the indicator component and
 * (later) the briefing. The loader that turns Supabase rows into these inputs
 * is `loadPostureInputs` in `./data`.
 */

export type Posture = 'green' | 'amber' | 'red'

/**
 * A regulatory deadline's bearing on posture: its severity and how many days
 * remain. `daysRemaining < 0` means the deadline is overdue.
 */
export interface PostureDeadline {
  severity: FindingSeverity
  daysRemaining: number
}

export interface PostureInputs {
  /** Severities of every OPEN (pending) finding. */
  openSeverities: FindingSeverity[]
  /** Approaching or overdue regulatory deadlines. */
  deadlines: PostureDeadline[]
}

/** The "near-term" window the AC pins posture to. */
export const NEAR_TERM_DAYS = 30

/**
 * Collapse the open findings and deadlines into one band.
 *
 *   * Red   — a Critical finding is open, or a Critical deadline is overdue.
 *   * Amber — a High finding is open, or a Critical/High deadline falls inside
 *             the 30-day window. (A near-term Critical deadline that isn't yet
 *             overdue still breaks Green per the AC, but it is not Red until it
 *             actually lapses — so it lands here.)
 *   * Green — nothing Critical/High is pressing.
 *
 * Precedence is Red → Amber → Green: the worst applicable band wins.
 */
export function computePosture({ openSeverities, deadlines }: PostureInputs): Posture {
  const hasOpen = (s: FindingSeverity) => openSeverities.includes(s)
  const deadlineOverdue = (s: FindingSeverity) =>
    deadlines.some((d) => d.severity === s && d.daysRemaining < 0)
  const deadlineNearTerm = (s: FindingSeverity) =>
    deadlines.some(
      (d) => d.severity === s && d.daysRemaining >= 0 && d.daysRemaining <= NEAR_TERM_DAYS,
    )

  if (hasOpen('critical') || deadlineOverdue('critical')) return 'red'

  if (hasOpen('high') || deadlineNearTerm('high') || deadlineNearTerm('critical')) return 'amber'

  return 'green'
}

export interface PostureMeta {
  posture: Posture
  /** The band name (AC: "Green / Amber / Red"). */
  label: string
  /** Plain-language summary shown next to the indicator. No jargon. */
  headline: string
  /** Tailwind classes for the large indicator dot, readable on the dark console. */
  dotClassName: string
  /** Text colour matching the band. */
  textClassName: string
}

const POSTURE_META: Record<Posture, PostureMeta> = {
  green: {
    posture: 'green',
    label: 'Green',
    headline: "You're on track. Nothing urgent right now.",
    dotClassName: 'bg-emerald-400 shadow-[0_0_60px_-5px] shadow-emerald-400/60',
    textClassName: 'text-emerald-300',
  },
  amber: {
    posture: 'amber',
    label: 'Amber',
    headline: 'Needs attention. A few things are coming due.',
    dotClassName: 'bg-amber-400 shadow-[0_0_60px_-5px] shadow-amber-400/60',
    textClassName: 'text-amber-300',
  },
  red: {
    posture: 'red',
    label: 'Red',
    headline: 'Action required. You have something critical open.',
    dotClassName: 'bg-rose-500 shadow-[0_0_60px_-5px] shadow-rose-500/60',
    textClassName: 'text-rose-300',
  },
}

export function postureMeta(posture: Posture): PostureMeta {
  return POSTURE_META[posture]
}

/** How the band is computed — surfaced as the indicator's tooltip (AC). */
export const POSTURE_TOOLTIP =
  'Green: no critical findings open and no critical or high deadlines within 30 days. ' +
  'Amber: an open high finding, or a critical/high deadline within 30 days. ' +
  'Red: an open critical finding, or a critical deadline overdue.'

File: apps/core-api/internal/domain/findings/posture.go
// Package findings holds the domain rules for the feed and the dashboard.
//
// Pure functions over already-loaded data, with no database and no proto, for
// the same reason the legacy `lib/dashboard/posture.ts` was written that way:
// the posture rule is the acceptance criterion, and a rule that can only be
// exercised through a query is a rule nobody exhaustively tests.
package findings

// Posture is the dashboard's single band.
type Posture string

const (
	// PostureNotAssessed means the Watcher has never run for this organisation.
	//
	// ENT-161, and the reason this type has four values rather than the legacy
	// three. A posture derived by counting findings returns green before
	// anything has looked, so the old console told founders "You're on track"
	// about a compliance record the product had never examined. For a product
	// whose value is that a human can check a claim, that is the worst
	// available failure: not a wrong answer, a confident one.
	//
	// It is first in the constant block and handled first in every switch
	// because it is not a severity band at all. It is the absence of an
	// assessment, and it outranks the others rather than sitting between them.
	PostureNotAssessed Posture = "not_assessed"

	PostureGreen Posture = "green"
	PostureAmber Posture = "amber"
	PostureRed   Posture = "red"
)

// NearTermDays is the window an approaching deadline has to fall inside to
// affect posture. Carried over from the legacy rule unchanged.
const NearTermDays = 30

// Deadline is an approaching or overdue regulatory deadline.
//
// DaysRemaining is negative when the deadline has passed.
//
// In the legacy stack these came from their own loader. There is no deadlines
// table in this schema and there never was: `watcher_detect_deadlines` emits
// them as signals, so by the time they reach here they are findings whose
// `metadata ->> 'signal_kind'` is `deadline` and whose signal metadata carries
// `days_remaining`. The rule is unchanged; only where the rows come from is.
type Deadline struct {
	Severity      string
	DaysRemaining int
}

// PostureInputs is everything the band is computed from.
type PostureInputs struct {
	// Severities of every open (pending) finding.
	OpenSeverities []string
	// Approaching or overdue deadlines.
	Deadlines []Deadline
	// Whether the Watcher has ever run for this organisation. False makes the
	// band NotAssessed whatever the other fields say, because they are then the
	// absence of evidence rather than evidence of absence.
	Assessed bool
}

// ComputePosture collapses open findings and deadlines into one band.
//
// The rule is the legacy one, reviewed at db0bf83 before being rewritten as
// ENT-200 requires, with ENT-161's fourth state added in front of it:
//
//   - NotAssessed — the Watcher has never run.
//   - Red         — a critical finding is open, or a critical deadline is overdue.
//   - Amber       — a high finding is open, or a critical/high deadline falls
//     inside the 30-day window.
//   - Green       — nothing critical or high is pressing.
//
// Precedence is NotAssessed, then Red, then Amber, then Green: the worst
// applicable band wins.
//
// A near-term critical deadline that has not yet lapsed is deliberately Amber
// rather than Red. It breaks Green because it is pressing, and it is not Red
// until it actually passes, which is the distinction the original AC drew and
// the one a founder acts on differently.
func ComputePosture(in PostureInputs) Posture {
	if !in.Assessed {
		return PostureNotAssessed
	}

	if hasSeverity(in.OpenSeverities, "critical") || overdue(in.Deadlines, "critical") {
		return PostureRed
	}

	if hasSeverity(in.OpenSeverities, "high") ||
		nearTerm(in.Deadlines, "high") ||
		nearTerm(in.Deadlines, "critical") {
		return PostureAmber
	}

	return PostureGreen
}

// Headline is the plain-language sentence shown with the band.
//
// Server-side so the wording is one thing rather than one per client, and
// carried over verbatim from the legacy console for the three bands that
// existed there. Changing them is a product decision, not a porting one.
func Headline(p Posture) string {
	switch p {
	case PostureNotAssessed:
		// Deliberately not reassuring. The honest content is that nothing has
		// been checked yet, and a sentence that sounds like good news here is
		// the exact bug ENT-161 describes.
		return "Not assessed yet. The Watcher has not run for this organisation."
	case PostureRed:
		return "Action required. You have something critical open."
	case PostureAmber:
		return "Needs attention. A few things are coming due."
	case PostureGreen:
		return "You're on track. Nothing urgent right now."
	default:
		// Unreachable through ComputePosture. Returning an empty string rather
		// than a cheerful default, because a band this function does not
		// recognise must not be rendered as if it were fine.
		return ""
	}
}

// SeverityCounts is the open-finding tally the stat row renders.
type SeverityCounts struct {
	Critical int
	High     int
	Medium   int
	Low      int
}

// Total is every open finding across all four severities.
func (c SeverityCounts) Total() int {
	return c.Critical + c.High + c.Medium + c.Low
}

// CountSeverities tallies open findings by severity.
//
// A severity outside the four known values is counted in none of them and is
// still absent from Total, which is deliberate: the column is check-constrained,
// so an unknown value means the constraint changed without this function, and
// silently folding it into "low" would under-report a customer's exposure.
func CountSeverities(severities []string) SeverityCounts {
	var c SeverityCounts
	for _, s := range severities {
		switch s {
		case "critical":
			c.Critical++
		case "high":
			c.High++
		case "medium":
			c.Medium++
		case "low":
			c.Low++
		}
	}
	return c
}

func hasSeverity(severities []string, want string) bool {
	for _, s := range severities {
		if s == want {
			return true
		}
	}
	return false
}

func overdue(deadlines []Deadline, severity string) bool {
	for _, d := range deadlines {
		if d.Severity == severity && d.DaysRemaining < 0 {
			return true
		}
	}
	return false
}

func nearTerm(deadlines []Deadline, severity string) bool {
	for _, d := range deadlines {
		if d.Severity == severity && d.DaysRemaining >= 0 && d.DaysRemaining <= NearTermDays {
			return true
		}
	}
	return false
}
